tools/bpf_inspect.py

```diff
@@ -0,0 +1,151 @@
+    # though so no error checking here.
+    BpfProgTrampType = enum_type_to_class(
+        prog.type("enum bpf_tramp_prog_type"), "BpfProgTrampType"
+    )
+
+    at = BpfAttachType(attach_type)
+
+    if at == BpfAttachType.BPF_TRACE_FENTRY:
+        return BpfProgTrampType.BPF_TRAMP_FENTRY
+
+    if at == BpfAttachType.BPF_TRACE_FEXIT:
+        return BpfProgTrampType.BPF_TRAMP_FEXIT
+
+    return BpfProgTrampType.BPF_TRAMP_REPLACE
+
+
+def get_linked_func(bpf_prog):
+    kind = attach_type_to_tramp(bpf_prog.expected_attach_type)
+
+    linked_prog = bpf_prog.aux.linked_prog
+    linked_prog_id = linked_prog.aux.id.value_()
+    linked_btf_id = bpf_prog.aux.attach_btf_id.value_()
+    linked_name = (
+        f"{get_prog_name(linked_prog)}->"
+        f"{get_btf_name(linked_prog.aux.btf, linked_btf_id)}()"
+    )
+
+    return f"{linked_prog_id}->{linked_btf_id}: {kind.name} {linked_name}"
+
+
+def get_tramp_progs(bpf_prog):
+    try:
+        tr = bpf_prog.aux.member_("trampoline")
+    except LookupError:
+        # Trampoline is available since Linux kernel commit
+        # fec56f5890d9 ("bpf: Introduce BPF trampoline") (in v5.5).
+        # Skip trampoline if current kernel doesn't support it.
+        return
+
+    if not tr:
+        return
+
+    if tr.extension_prog:
+        yield tr.extension_prog
+    else:
+        for head in tr.progs_hlist:
+            for tramp_aux in hlist_for_each_entry(
+                "struct bpf_prog_aux", head, "tramp_hlist"
+            ):
+                yield tramp_aux.prog
+
+
+def list_bpf_progs(args, prog):
+    for bpf_prog in bpf_prog_for_each(prog):
+        id_ = bpf_prog.aux.id.value_()
+        type_ = BpfProgType(bpf_prog.type).name
+        name = get_prog_name(bpf_prog)
+
+        linked = ", ".join(
+            [get_linked_func(p) for p in get_tramp_progs(bpf_prog)]
+        )
+        if linked:
+            linked = f" linked:[{linked}]"
+
+        print(f"{id_:>6}: {type_:32} {name:32} {linked}")
+
+
+def list_bpf_maps(args, prog):
+    for map_ in bpf_map_for_each(prog):
+        id_ = map_.id.value_()
+        type_ = BpfMapType(map_.map_type).name
+        name = map_.name.string_().decode()
+
+        print(f"{id_:>6}: {type_:32} {name}")
+
+
+def main():
+    parser = argparse.ArgumentParser(
+        description=DESCRIPTION, formatter_class=argparse.RawTextHelpFormatter
+    )
+
+    subparsers = parser.add_subparsers(
+        title="subcommands", dest="subcommand", required=True
+    )
+
+    prog_parser = subparsers.add_parser(
+        "prog", aliases=["p"], help="list BPF programs"
+    )
+    prog_parser.set_defaults(func=list_bpf_progs)
+
+    map_parser = subparsers.add_parser(
+        "map", aliases=["m"], help="list BPF maps"
+    )
+    map_parser.set_defaults(func=list_bpf_maps)
+
+    args = parser.parse_args()
+    args.func(args, prog)  # prog is global drgn.Program provided by drgn.
```

Comment:
We don't need to pass `prog` around since it's global, right?

Reply:
Right, it was my poor try to highlight that it's not a good idea to use `prog` as a variable for bpf prog e.g. in `list_bpf_progs` (what was very tempting since `prog` is what is used in kernel all over the place for `struct bpf_prog`).
But I guess it doesn't add much value and only adds confusion as in this case, fine to remove as well.